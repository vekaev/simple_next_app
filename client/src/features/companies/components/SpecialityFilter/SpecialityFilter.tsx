import React, { useCallback, useMemo } from 'react';
import { SimpleGrid } from '@mantine/core';

import { TitledCheckbox } from '@components';
import { useCompanies } from '../../hooks/useCompanies';

const SpecialityFilter: React.FC = () => {
  const { specialties, filters, setFilters } = useCompanies();

  const activeSpecialtiesIds = useMemo(
    () =>
      specialties
        .map(specialty => specialty.id)
        .filter(id => filters.specialties.includes(id)),
    [specialties, filters.specialties]
  );
  const handleCheckboxChange = useCallback(
    (specialtyId: string) => (isChecked: boolean) => {
      if (isChecked) {
        setFilters({
          ...filters,
          specialties: [...filters.specialties, specialtyId],
        });
      } else {
        setFilters({
          ...filters,
          specialties: filters.specialties.filter(
            specialtyIdToCheck => specialtyIdToCheck !== specialtyId
          ),
        });
      }
    },
    [filters]
  );

  return (
    <SimpleGrid
      cols={3}
      breakpoints={[
        { maxWidth: 'md', cols: 2 },
        { maxWidth: 'sm', cols: 1 },
      ]}
    >
      {specialties.map(specialty => (
        <TitledCheckbox
          key={specialty.id}
          title={specialty.name}
          checked={activeSpecialtiesIds.includes(specialty.id)}
          onChange={handleCheckboxChange(specialty.id)}
        />
      ))}
    </SimpleGrid>
  );
};

export default SpecialityFilter;
