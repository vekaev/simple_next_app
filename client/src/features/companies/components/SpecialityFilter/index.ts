export { default as SpecialityFilter } from './SpecialityFilter';
