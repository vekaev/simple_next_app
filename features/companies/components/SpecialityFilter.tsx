import React, { useCallback } from 'react';
import { SimpleGrid } from '@mantine/core';

import { TitledCheckbox } from '@components';
import { useCompanies } from '../hooks/useCompanies';

const SpecialityFilter: React.FC = () => {
  const { specialtiesList, filters, setCompaniesFilters } = useCompanies();

  const handleCheckboxChange = useCallback(
    (specialty: string) => (isChecked: boolean) => {
      if (isChecked) {
        setCompaniesFilters({
          specialties: [...filters.specialties, specialty],
        });
      } else {
        setCompaniesFilters({
          specialties: filters.specialties.filter(
            specialtyToCheck => specialtyToCheck !== specialty
          ),
        });
      }
    },
    [filters.specialties]
  );

  return specialtiesList.length ? (
    <SimpleGrid
      cols={4}
      breakpoints={[
        { maxWidth: 'md', cols: 2 },
        { maxWidth: 'sm', cols: 1 },
      ]}
    >
      {specialtiesList.map(name => (
        <TitledCheckbox
          key={name}
          title={name}
          checked={!!filters.specialties?.includes(name)}
          onChange={handleCheckboxChange(name)}
        />
      ))}
    </SimpleGrid>
  ) : null;
};

export default SpecialityFilter;
