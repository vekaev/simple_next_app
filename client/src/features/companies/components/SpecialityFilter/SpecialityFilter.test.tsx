import { screen } from '@testing-library/react';
import { renderWithCompanyProvider } from '@features/companies/utils/test/renderWithCompanyProvider';

import SpecialityFilter from './SpecialityFilter';

describe('SpecialityFilter', () => {
  test('renders correctly', async () => {
    const { container } = renderWithCompanyProvider(<SpecialityFilter />);

    const specialityFilters = await screen.findAllByRole('checkbox');
    const specialityFilterTitles = specialityFilters.map(
      specialityFilter => specialityFilter.title
    );

    expect(specialityFilters).toHaveLength(3);
    specialityFilters.forEach(specialityFilter =>
      expect(specialityFilter).not.toBeChecked()
    );
    expect(specialityFilterTitles).toEqual([
      'Excavation',
      'Plumbing',
      'Electrical',
    ]);

    expect(container).toMatchSnapshot();
  });
});
