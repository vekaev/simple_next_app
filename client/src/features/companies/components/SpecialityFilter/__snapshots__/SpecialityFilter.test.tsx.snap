// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpecialityFilter renders correctly 1`] = `
<div>
  <div
    class="mantine-SimpleGrid-root mantine-akbhe9"
  >
    <div
      class="mantine-u85prc"
    >
      <div
        class="mantine-s7fa4k"
      >
        <div
          class="mantine-Text-root mantine-1w7xm3i"
        >
          Excavation
        </div>
      </div>
      <div
        class="mantine-Checkbox-root mantine-16ttirm"
      >
        <div
          class="mantine-qbjb2s mantine-Checkbox-inner"
        >
          <input
            class="mantine-sov5oy mantine-Checkbox-input"
            id="mantine-r0"
            tabindex="-1"
            title="Excavation"
            type="checkbox"
          />
          <svg
            class="__mantine-ref-icon mantine-180iq9w mantine-Checkbox-icon"
            fill="none"
            viewBox="0 0 10 7"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M4 4.586L1.707 2.293A1 1 0 1 0 .293 3.707l3 3a.997.997 0 0 0 1.414 0l5-5A1 1 0 1 0 8.293.293L4 4.586z"
              fill="currentColor"
              fill-rule="evenodd"
            />
          </svg>
        </div>
      </div>
    </div>
    <div
      class="mantine-u85prc"
    >
      <div
        class="mantine-s7fa4k"
      >
        <div
          class="mantine-Text-root mantine-1w7xm3i"
        >
          Plumbing
        </div>
      </div>
      <div
        class="mantine-Checkbox-root mantine-16ttirm"
      >
        <div
          class="mantine-qbjb2s mantine-Checkbox-inner"
        >
          <input
            class="mantine-sov5oy mantine-Checkbox-input"
            id="mantine-r1"
            tabindex="-1"
            title="Plumbing"
            type="checkbox"
          />
          <svg
            class="__mantine-ref-icon mantine-180iq9w mantine-Checkbox-icon"
            fill="none"
            viewBox="0 0 10 7"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M4 4.586L1.707 2.293A1 1 0 1 0 .293 3.707l3 3a.997.997 0 0 0 1.414 0l5-5A1 1 0 1 0 8.293.293L4 4.586z"
              fill="currentColor"
              fill-rule="evenodd"
            />
          </svg>
        </div>
      </div>
    </div>
    <div
      class="mantine-u85prc"
    >
      <div
        class="mantine-s7fa4k"
      >
        <div
          class="mantine-Text-root mantine-1w7xm3i"
        >
          Electrical
        </div>
      </div>
      <div
        class="mantine-Checkbox-root mantine-16ttirm"
      >
        <div
          class="mantine-qbjb2s mantine-Checkbox-inner"
        >
          <input
            class="mantine-sov5oy mantine-Checkbox-input"
            id="mantine-r2"
            tabindex="-1"
            title="Electrical"
            type="checkbox"
          />
          <svg
            class="__mantine-ref-icon mantine-180iq9w mantine-Checkbox-icon"
            fill="none"
            viewBox="0 0 10 7"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M4 4.586L1.707 2.293A1 1 0 1 0 .293 3.707l3 3a.997.997 0 0 0 1.414 0l5-5A1 1 0 1 0 8.293.293L4 4.586z"
              fill="currentColor"
              fill-rule="evenodd"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
`;
